### DISCUSSION TITLE
What is correct parse_qsl or parse_qs?

### DISCUSSION THREAD

**Discussion by ASamoylov** [Kludex/starlette#1949] [Q&A]
https://github.com/encode/starlette/blob/bf5d7357b544b48b96844d3274ea558df079571d/starlette/datastructures.py#L407

```
What if I want to use query parameter list / multiple values?
Check please

from urllib.parse import SplitResult, parse_qsl, parse_qs
from starlette.datastructures import QueryParams, ImmutableMultiDict
qs = "q=foo&q=bar"
parse_qsl(qs)  # [('q', 'foo'), ('q', 'bar')]

parse_qs(qs)  # {'q': ['foo', 'bar']}

QueryParams(qs)._dict  # {'q': 'bar'}

QueryParams(qs).getlist("q")  # ['foo', 'bar']
```

**Comment by vladpereverzyev** [comment]
`parse_qsl` and `parse_qs` are both correct, but they return different representations.

For a query string like:

```python
qs = "q=foo&q=bar"
```

* `parse_qsl(qs)` returns a list of key/value pairs:

  ```python
  [('q', 'foo'), ('q', 'bar')]
  ```

* `parse_qs(qs)` groups repeated keys into lists:

  ```python
  {'q': ['foo', 'bar']}
  ```

For Starlette, `QueryParams` is designed to preserve multiple values as well:

```python
params = QueryParams(qs)

params.getlist("q")
# ['foo', 'bar']
```

So if you need all values for a repeated query parameter, `getlist()` is the appropriate API.

I would avoid relying on `QueryParams._dict`, since `_dict` is an internal implementation detail and is not the public API for accessing query parameters.

In other words, there is no issue with Starlette using `parse_qsl` here: `QueryParams` provides the multi-value behavior through `getlist()`.
